File: ui/src/api/local-payment.ts
import { loggedInvoke } from '@/utils/logged-invoke';

// ── Local payment methods (regional slice 6) ────────────────────────

/**
 * Who supplied an effective payment rail (mirrors
 * `kasirmu_core::regional::ConfigScope`'s serde names verbatim).
 */
export type PaymentRailScope = 'location' | 'legal_entity';

/**
 * One effective payment rail for a location: the MARKET surface — which
 * rails exist, what they are called, whether the site offers them.
 *
 * Deliberately carries NO tier answer: `supports_qris` is a tier capability
 * from the license layer, never derivable from these rows, and no gateway
 * credential can appear in `parameters` (the backend write path rejects
 * credential-shaped keys).
 */
export interface LocalPaymentRail {
  /** Stable rail code (e.g. `qris`, `va-bca`). */
  rail_code: string;
  /** Display label. */
  label: string;
  /** Whether the site offers the rail. */
  is_enabled: boolean;
  /** Who answered: the market default (legal entity) or the site. */
  scope: PaymentRailScope;
  /** Per-rail market metadata (JSON object string). Never credentials. */
  parameters: string;
}

/** One rail in the card's replace-set submission. */
export interface LocalPaymentRailArgs {
  /** Stable rail code. */
  rail_code: string;
  /** Display label. */
  label: string;
  /** Whether the scope offers the rail. */
  is_enabled: boolean;
  /** Per-rail market metadata (JSON object string). Never credentials. */
  parameters: string;
}

/** Get the effective payment rails for one location (ADR #7 scoped,
 *  gated `settings:read` server-side). */
export const getLocalPaymentMethodsScoped = (
  sessionToken: string,
  locationId: string,
): Promise<LocalPaymentRail[]> =>
  loggedInvoke<LocalPaymentRail[]>('get_local_payment_methods_scoped', {
    sessionToken,
    locationId,
  });

/** Replace the location's rail list (the card's whole-list write) and get
 *  the freshly effective surface back (gated `settings:edit` server-side). */
export const setLocalPaymentMethodsScoped = (
  sessionToken: string,
  locationId: string,
  rails: LocalPaymentRailArgs[],
): Promise<LocalPaymentRail[]> =>
  loggedInvoke<LocalPaymentRail[]>('set_local_payment_methods_scoped', {
    sessionToken,
    locationId,
    rails,
  });

// ── The qris rail's static-QR payload (agents-5 R2) ─────────────────
//
// Key inside the per-rail `parameters` bag carrying the merchant's
// static EMVCo QRIS string. It is market metadata — the same code the
// printed counter poster encodes, scannable by anyone — never a
// credential, which is the distinction the backend's
// FORBIDDEN_PARAMETER_FRAGMENTS guard enforces on this exact bag.

export const STATIC_QR_PARAM = 'static_qr_payload';

/** Read the payload from a rail's parameters JSON; malformed or absent
 *  reads as null ("not configured") — the fail-open truth. */
export function readStaticQrPayload(parameters: string): string | null {
  try {
    const parsed: unknown = JSON.parse(parameters);
    if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
      const value = (parsed as Record<string, unknown>)[STATIC_QR_PARAM];
      return typeof value === 'string' && value.length > 0 ? value : null;
    }
  } catch {
    /* malformed bag */
  }
  return null;
}

/** Write the payload into a parameters bag, preserving every other key
 *  (an empty value removes the key). A malformed incoming bag is NOT
 *  silently replaced — the backend rejects malformed bags too, so the
 *  first honest save must be a deliberate repair. */
export function writeStaticQrPayload(parameters: string, value: string): string {
  let parsed: Record<string, unknown>;
  try {
    const raw: unknown = JSON.parse(parameters);
    if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return parameters;
    parsed = raw as Record<string, unknown>;
  } catch {
    return parameters;
  }
  if (value.length === 0) delete parsed[STATIC_QR_PARAM];
  else parsed[STATIC_QR_PARAM] = value;
  return JSON.stringify(parsed);
}

File: apps/desktop-client/src/commands/local_payment.rs
//! Local payment method commands (regional slice 6, saas-2 design slice
//! queue #6) — the market rail surface, entity → location.
//!
//! Wire contract: the core `EffectivePaymentRail` model is returned as-is
//! (camelCase serde — the shape the dev-mock and the card mirror). The
//! module never touches entitlements or `payment_gateways`: which rails a
//! market/site offers is a MARKET fact; `supports_qris` is a TIER answer
//! (license layer) and gateway credentials live in `payment_gateways`. The
//! write path rejects credential-shaped parameter keys outright.
//!
//! Read: `settings:read` (any operator can see the rail surface).
//! Write: `settings:edit` + the ADR #47 location-resource gate — the
//! location layer is what the card edits; the legal-entity layer is
//! managed through the same command with the entity scope resolved by
//! later management surfaces.

use oz_core::db::assignments::ScopeType;
use oz_core::db::payment_methods::{EffectivePaymentRail, NewPaymentRail};
use oz_core::{Store, permissions};
use tauri::State;

use crate::commands::authz::{
    require_permission_for_session, require_permission_for_session_resource,
};
use crate::error::AppError;
use crate::state::AppState;

/// One rail in the card's replace-set submission.
#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalPaymentRailArgs {
    /// Stable rail code (e.g. `qris`, `va-bca`).
    pub rail_code: String,
    /// Display label.
    pub label: String,
    /// Whether the scope offers the rail.
    pub is_enabled: bool,
    /// Per-rail market metadata (JSON object). Credential-shaped keys are
    /// rejected by the core write path.
    #[serde(default)]
    pub parameters: String,
}

/// Read the effective payment-rail surface for one location of the
/// session's store (regional slice 6).
///
/// Resolves the session's store database (ADR #7) and checks `settings:read`
/// inline. The read walks entity → location per rail with provenance — the
/// same chain semantics as the regional read. An unlinked or unknown
/// location answers an empty list (no market, no invented rails).
#[tauri::command]
pub async fn get_local_payment_methods_scoped(
    location_id: String,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<EffectivePaymentRail>, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_permission_for_session(&state, &session, permissions::SETTINGS_READ).await?;
    let conn = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&conn);
    Ok(store.local_payment_methods_for_location(&location_id)?)
}

/// Replace the location's rail list (the card's whole-list write,
/// transactional) and return the freshly effective surface (read-after-
/// write, same connection) so the card re-renders provenance without a
/// second round-trip.
///
/// Checks `settings:edit` scoped to the location resource (ADR #47).
/// Validation happens in core — credential-shaped parameter keys, blank
/// codes/labels, duplicates — inside the same transaction that rewrites
/// the rows; this command adds no validation of its own.
#[tauri::command]
pub async fn set_local_payment_methods_scoped(
    location_id: String,
    rails: Vec<LocalPaymentRailArgs>,
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<EffectivePaymentRail>, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_permission_for_session(&state, &session, permissions::SETTINGS_EDIT).await?;
    require_permission_for_session_resource(
        &state,
        &session,
        permissions::SETTINGS_EDIT,
        ScopeType::Location,
        &location_id,
    )
    .await?;
    let submitted: Vec<NewPaymentRail> = rails
        .into_iter()
        .map(|r| NewPaymentRail {
            rail_code: r.rail_code,
            label: r.label,
            is_enabled: r.is_enabled,
            parameters: r.parameters,
        })
        .collect();
    let conn = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&conn);
    let now = chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
    store.replace_local_payment_methods("location", &location_id, &submitted, &now)?;
    Ok(store.local_payment_methods_for_location(&location_id)?)
}

#[cfg(test)]
#[path = "local_payment_tests.rs"]
mod tests;
